Attempt to document design and source layout

evs-chris opened the pull request:
I cleaned _at_ some of my notes and popped them into a readme that tries to give a general overview of the major parts of Ractive, what they do, and how they're arranged in the src tree. I'm not at all a writer of words, so it's my usual rambley run-on text, but hopefully it makes it a little easier to understand what Ractive does under the hood. I hope to add a few more detailed breakdowns in some of the other child directories soon, but I thought I'd throw this into a PR/branch for feedback and possibly some collaboration.

So, feedback and pushes welcome! :smile:

fskreuz commented:
Referencing the other issues related to internals:

- https://github.com/ractivejs/ractive/issues/1972
- https://github.com/ractivejs/docs.ractivejs.org/issues/47
- https://github.com/ractivejs/ractive/issues/2363
- https://github.com/ractivejs/ractive/issues/2411

fskreuz commented:
A good things to touch in an overview would be:

- How the viewmodel, virtual DOM and the actual DOM are connected, who listens/calls what, etc.
- How the viewmodel looks.
- How the vdom looks.
- How the DOM is constructed/reconstructed.
- How the "surgical" changes are done.
- The things that happen when data changes happen, like if a `set` is called.
- The things that happen when one toggles an `{{#if}}` block.
- How the runloop works, what happens when.

I believe a visual representation is a good way to explain the flow, and everything will fall into place once the basic flow is understood. Thinking ASCII art would do the trick.

evs-chris commented:
I split the main README into three additional READMEs and added a fair amount of slightly more coherent info :smile:. I haven't found a decent ascii art flow charter or the willingness to flow chart :wink:, but I'll see if I can tackle that later. I'mma merge this now and address anything else later. Feel free to drop more requests, comments, and corrections in this thread though, as I'll keep an eye on it.
